pkg/flow/events.go

```diff
@@ -1005,3 +1015,161 @@ func (events *events) listenForEvents(ctx context.Context, im *instanceMemory, c
+		return nil, err
+	}
+
+	if script, ok := cache[key]; ok {
+		SCRIPT = script
+
+	} else {
+		ceventfilter, err := ns.QueryCloudeventfilters().Where(enteventsfilter.NameEQ(filtername)).Only(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		SCRIPT = ceventfilter.Jscode
+
+	}
+
+	var mapEvent map[string]interface{}
+	err = json.Unmarshal(cloudevent, &mapEvent)
+	if err != nil {
+		return nil, err
+	}
+
+	//create js runtime
+	vm := goja.New()
+
+	vm.Set("event", mapEvent)
+
+	_, err = vm.RunString(SCRIPT)
+	if err != nil {
+		panic(err)
+	}
+
+	var fn func() map[string]interface{}
+	err = vm.ExportTo(vm.Get("filter"), &fn)
+	if err != nil {
+		panic(err)
+	}
+
+	newEventMap := fn()
+	newBytesEvent, err := json.Marshal(newEventMap)
+	if err != nil {
+		return nil, err
+	}
+
+	resp := new(grpc.ApplyCloudEventFilterResponse)
+
+	resp.Event = newBytesEvent
+
+	return resp, nil
+}
+
+func (flow *flow) DeleteCloudEventFilter(ctx context.Context, in *grpc.DeleteCloudEventFilterRequest) (*emptypb.Empty, error) {
+	//delete filtername from database
+
+	namespace := in.GetNamespace()
+	filtername := in.GetFilterName()
+
+	ns, err := flow.getNamespace(ctx, flow.db.Namespace, namespace)
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = ns.QueryCloudeventfilters().Where(enteventsfilter.NameEQ(filtername)).Only(ctx)
+	if err != nil {
+		err = errors.New("event filter not existing")
+		return nil, err
+	}
+
+	_, err = flow.db.CloudEventFilters.
+		Delete().
+		Where(
+			enteventsfilter.And(
+				enteventsfilter.NameEQ(filtername),
+			)).
+		Exec(context.Background())
+
+	if err != nil {
+		return nil, err
+	}
+
+	var key keyPair
+	key.filtername = filtername
+	key.namespace = namespace
+	delete(cache, key)
+
+	var resp emptypb.Empty
+
+	return &resp, err
+
+}
+
+func (flow *flow) CreateCloudEventFilter(ctx context.Context, in *grpc.CreateCloudEventFilterRequest) (*emptypb.Empty, error) {
+	//TODO
+	//get per filtername to check if its already existing
+	//create filter in databse with im.filtername & im.jsCode
+
+	namespace := in.GetNamespace()
+	filtername := in.GetFiltername()
+	SCRIPT := in.GetJsCode()
+
+	SCRIPT = fmt.Sprintf("function filter() {\n %s \n}", SCRIPT)
+
+	//compiling js code is needed
+	_, err := goja.Compile("filter", SCRIPT, false)
+	if err != nil {
+		err := errors.New("js code does not compile")
+		return nil, err
+	}
+
+	ns, err := flow.getNamespace(ctx, flow.db.Namespace, namespace)
+	if err != nil {
+		return nil, err
+	}
+
+	ceventfilter, _ := ns.QueryCloudeventfilters().Where(enteventsfilter.NameEQ(filtername)).Only(ctx)
+	if ceventfilter != nil {
+		err = errors.New("event filter already exists")
+		return nil, err
+	}
+
+	_, err = flow.db.CloudEventFilters.Create().SetName(filtername).SetNamespace(ns).SetJscode(SCRIPT).Save(context.Background())
+	if err != nil {
+		return nil, err
+	}
+
+	var key keyPair
+	key.filtername = filtername
+	key.namespace = namespace
+	cache[key] = SCRIPT
+
+	var resp emptypb.Empty
```

Comment:
for consistency I would create the var resp at the beginning and return resp, err every time.